Core/include/Acts/EventData/SingleBoundTrackParameters.hpp

```diff
@@ -91,22 +92,27 @@ class SingleBoundTrackParameters {
   /// @param p Absolute momentum
   /// @param q Particle charge
   /// @param cov Bound parameters covariance matrix
-  SingleBoundTrackParameters(std::shared_ptr<const Surface> surface,
-                             const GeometryContext& geoCtx, const Vector4& pos4,
-                             const Vector3& dir, Scalar p, Scalar q,
-                             std::optional<CovarianceMatrix> cov = std::nullopt)
-      : m_params(detail::transformFreeToBoundParameters(
-            pos4.segment<3>(ePos0), pos4[eTime], dir,
-            (q != Scalar(0)) ? (q / p) : (1 / p), *surface, geoCtx)),
-        m_cov(std::move(cov)),
-        m_surface(std::move(surface)),
-        m_chargeInterpreter(std::abs(q)) {
-    assert((0 <= p) and "Absolute momentum must be positive");
-    assert(m_surface);
-    // free-to-bound transform always return phi/theta within bounds
+  ///
+  /// @note The returned result indicates whether the free parameters could
+  /// successfully be converted to on-surface parameters.
+  static Result<SingleBoundTrackParameters<charge_t>> create(
+      std::shared_ptr<const Surface> surface, const GeometryContext& geoCtx,
+      const Vector4& pos4, const Vector3& dir, Scalar p, Scalar q,
+      std::optional<CovarianceMatrix> cov = std::nullopt) {
+    Result<BoundVector> bound = detail::transformFreeToBoundParameters(
+        pos4.segment<3>(ePos0), pos4[eTime], dir,
+        (q != Scalar(0)) ? (q / p) : (1 / p), *surface, geoCtx);
+
+    if (!bound.ok()) {
+      return bound.error();
+    }
+
+    return SingleBoundTrackParameters<charge_t>{std::move(surface), *bound, q,
+                                                std::move(cov)};
```

Comment:
This is a big change in philosophy though, meaning we give up on object creation itself in favour of factories for certain cases where garbled input would create a useless object. I think it's fine as long as we keep the unchecked constructors that are local + surface, because they should not pay a penality through checking.

Reply:
Ha, that's why I put the TODO in, so I don't forget 😅

Reply:
Yeah that's what I did. There's no way around this though, because a constructor can't return anything. We would then have to use exceptions, which I think we decided not to in favor of `Result` a while ago.

Reply:
ok.